runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -0,0 +1,1123 @@
+function M.get_virtual_text_chunks_for_line(bufnr, line, line_diagnostics)
+  assert(bufnr and line)
+
+  if #line_diagnostics == 0 then
+    return nil
+  end
+
+  -- Create a little more space between virtual text and contents
+  local virt_texts = {{"  "}}
+
+  for i = 1, #line_diagnostics - 1 do
+    table.insert(virt_texts, {"■", virtual_text_highlight_map[line_diagnostics[i].severity]})
+  end
+  local last = line_diagnostics[#line_diagnostics]
+  -- TODO(ashkan) use first line instead of subbing 2 spaces?
+
+  -- TODO(tjdevries): Allow different servers to be shown first somehow?
+  -- TODO(tjdevries): Display server name associated with these?
+  if last.message then
+    table.insert(
+      virt_texts,
+      {
+        string.format("■ %s", last.message:gsub("\r", ""):gsub("\n", "  ")),
+        virtual_text_highlight_map[last.severity]
+      }
+    )
+
+    return virt_texts
+  end
+end
+-- }}}
+-- }}}
+-- Diagnostic Clear {{{
+--- Clears the currently displayed diagnostics
+---@param bufnr number The buffer number
+---@param client_id number the client id
+---@param diagnostic_ns number|nil Associated diagnostic namespace
+---@param sign_ns number|nil Associated sign namespace
+function M.clear(bufnr, client_id, diagnostic_ns, sign_ns)
+  validate { bufnr = { bufnr, 'n' } }
+
+  bufnr = (bufnr == 0 and api.nvim_get_current_buf()) or bufnr
+
+  if client_id == nil then
+    return vim.lsp.for_each_buffer_client(bufnr, function(_, iter_client_id, _)
+      return M.clear(bufnr, iter_client_id)
+    end)
+  end
+
+  diagnostic_ns = diagnostic_ns or M._get_diagnostic_namespace(client_id)
+  sign_ns = sign_ns or M._get_sign_namespace(client_id)
+
+  assert(bufnr, "bufnr is required")
+  assert(diagnostic_ns, "Need diagnostic_ns, got nil")
+  assert(sign_ns, string.format("Need sign_ns, got nil %s", sign_ns))
+
+  -- clear sign group
+  vim.fn.sign_unplace(sign_ns, {buffer=bufnr})
+
+  -- clear virtual text namespace
+  api.nvim_buf_clear_namespace(bufnr, diagnostic_ns, 0, -1)
+end
+-- }}}
+-- Diagnostic Insert Leave Handler {{{
+
+--- Callback scheduled for after leaving insert mode
+---
+--- Used to handle
+--@private
+function M._execute_scheduled_display(bufnr, client_id)
+  local args = table.remove(_bufs_waiting_to_update[bufnr], client_id)
+
+  M.display(nil, bufnr, client_id, args)
+end
+
+local registered = {}
+
+local make_augroup_key = function(bufnr, client_id)
+  return string.format("LspDiagnosticInsertLeave:%s:%s", bufnr, client_id)
+end
+
+--- Table of autocmd events to fire the update for displaying new diagnostic information
+M.insert_leave_auto_cmds = { "InsertLeave", "CursorHoldI" }
+
+--- Used to schedule diagnostic updates upon leaving insert mode.
+---
+--- For parameter description, see |M.display()|
+function M._schedule_display(bufnr, client_id, args)
+  if _bufs_waiting_to_update[bufnr][client_id] then
+    _bufs_waiting_to_update[bufnr][client_id] = args
+    return
+  end
+
```

Comment:
It looks like `_buf_waiting_to_update[bufnr][client_id]` never gets set to the args/config table, which causes `execute_scheduled_display` to call `display` with a `nil` config argument on line 767

It'll wipe out config provided with `vim.lsp.with` on scheduled display. I noticed this when virtual diagnostic text was getting enabled sporadically despite using `vim.lsp.with(<...>, { virtual_text = false })`

```suggestion
_bufs_waiting_to_update[bufnr][client_id] = args
```

Reply:
The above change fixes the bug when using sumneko_lua language server, but virtual text is still getting enabled when using multiple language servers (tsserver, diagnosticls), unsure what is causing it.

Maybe something to do with the the `InsertLeave,CursorHoldI` autocmd firing twice? Not sure how autocommands work in this respect

Reply:
Ah, great find. I added a different fix and then two tests to make sure this doesn't happen.